include/igl/embree/EmbreeRenderer.cpp

```diff
@@ -0,0 +1,457 @@
+#include "EmbreeRenderer.h"
+// Implementation
+
+//IGL viewing parts
+#include "../unproject.h"
+#include "../look_at.h"
+#include "../frustum.h"
+#include "../ortho.h"
+
+// color map
+#include "../jet.h"
+
+#include "../PI.h"
+
+// This unfortunately cannot be a static field of EmbreeRenderer because it
+// would depend on the template and then we might end up with initializing
+// embree twice. If only there was a way to ask embree if it's already
+// initialized...
+namespace igl
+{
+  namespace embree
+  {
+    // Keeps track of whether the **Global** Embree intersector has been
+    // initialized. This should never been done at the global scope.
+    static RTCDevice g_device = nullptr;
```

Comment:
We should use a singleton pattern to allocate/deallocate a single device shared by libigl code. For example something like that:

```cpp
class EmrbeeContext
{
public:
    static RTCDevice& device() {
        static EmrbeeContext context;
        return context.m_device;
    }

private:
    EmrbeeContext() {
        m_device = rtcNewDevice();
        if (rtcGetDeviceError (g_device) != RTC_ERROR_NONE) {
            std::cerr << "Embree: An error occurred while initializing embree core!" << std::endl;
        #ifdef IGL_VERBOSE
        } else {
            std::cerr << "Embree: core initialized." << std::endl;
        #endif
        }
    }

    ~RTCDeviceManager() {
        rtcReleaseDevice(m_device);
    }

    RTCDevice m_device;
};
```

Reply:
I created EmbreeDevice class with static functions to manage RTCDevice instance